File: 05_Abstraction/Interface/Main.java
interface Animal {
    void sound();
    void sleep();
}

class Dog implements Animal {
    @Override
    public void sound() {
        System.out.println("The dog barks.");
    }

    @Override
    public void sleep() {
        System.out.println("The dog sleeps.");
    }
}

class Cat implements Animal {
    @Override
    public void sound() {
        System.out.println("The cat meows.");
    }

    @Override
    public void sleep() {
        System.out.println("The cat sleeps.");
    }
}

public class Main {
    public static void main(String[] args) {
        Animal dog = new Dog();
        Animal cat = new Cat();

        dog.sound(); 
        dog.sleep();
        cat.sound(); 
        cat.sleep(); 
    }
}

File: 05_Abstraction/Interface/Main.cpp
#include <iostream>
using namespace std;

class Animal {
public:
    virtual void sound() = 0; 
    virtual void sleep() = 0; 
};

class Dog : public Animal {
public:
    void sound() override {
        cout << "The dog barks." << endl;
    }

    void sleep() override {
        cout << "The dog sleeps." << endl;
    }
};

class Cat : public Animal {
public:
    void sound() override {
        cout << "The cat meows." << endl;
    }

    void sleep() override {
        cout << "The cat sleeps." << endl;
    }
};

int main() {
    Animal* dog = new Dog();
    Animal* cat = new Cat();

    dog->sound(); 
    dog->sleep();
    cat->sound(); 
    cat->sleep();

    delete dog;
    delete cat;
    return 0;
}
